### DISCUSSION TITLE
Performance Difference

### DISCUSSION THREAD

**Discussion by tpatalas** [pmndrs/zustand#999] [Q&A] [answered]
Is there a performance difference between the `atom pick` vs `object pick`

Atomic Pick
```ts
const nuts = useStore(state => state.nuts)
const honey = useStore(state => state.honey)
```

vs

Object Pick
```ts
// Object pick, re-renders the component when either state.nuts or state.honey change
const { nuts, honey } = useStore(state => ({ nuts: state.nuts, honey: state.honey }))
```

**Comment by dai-shi** [comment, marked as answer]
```js
// "Object Pick"
const { nuts, honey } = useStore(state => ({ nuts: state.nuts, honey: state.honey }))
// That ☝️ creates a new object every time, so

// it behaves like this 👇 .
const { nuts, honey } = useStore()

// What you want instead is this 👇 .
const { nuts, honey } = useStore(state => ({ nuts: state.nuts, honey: state.honey }), shallow)
// Then, it's the same as "Atomic Pick" in terms of render optimization.
```

**Reply by tpatalas** [reply]
Wow..Thank you for the clarification and  appreciate for the amazing library!

**Reply by micahjon** [reply]
Thanks @dai-shi , this makes sense. 

If I understand correctly, in all 3 cases, the `nuts` and `honey` variables are identical (exact same value/reference), so the issue isn't triggering dependent hooks (e.g. `useEffect(..., [nuts])`. Instead, this optimization means that if another part of state (e.g. `state.apples`) is updated, then the component won't re-render, since shallow equality will check _only_ the `nuts` and `honey` keys.

Please let me know if I'm missing something!

Also, I'm curious if the `state` reference changes when `setState() / set()` is called. I'm assuming it does for `useStore()` to work (since the `state` reference would have to change for strict equality to force a re-render). 

Thanks! I've really enjoyed using Zustand and just want to be 100% sure I understand the details.

**Reply by dai-shi** [reply]
> this optimization means that if another part of state (e.g. `state.apples`) is updated, then the component won't re-render

correct.



> Also, I'm curious if the `state` reference changes when `setState() / set()` is called.

Yes, the `state` will be a new reference with `setState/set` because of `Object.assign({}, ...)`.

https://github.com/pmndrs/zustand/blob/197e5d41993b0e3ad6ebef9a7e6a9161753072f1/src/vanilla.ts#L94

**Reply by DarlonHenrique** [reply]
What is that "shallow" in the end of `const { nuts, honey } = useStore(state => ({ nuts: state.nuts, honey: state.honey }), shallow)` ??

**Reply by dai-shi** [reply]
It's a shallow function, which is exported from `zustand/shallow`, but it can be anything.
